util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: pp,
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for i := 0; i < ni.FieldValue.Len(); i++ {
+				nn := *ni
+				// The schema for each element is the list schema minus the list
+				// attrs.
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = pp
+				nn.FieldValue = ni.FieldValue.Index(i)
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 
-	case IsValueMap(ni.FieldValue):
-		for _, key := range ni.FieldValue.MapKeys() {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.MapIndex(key)
-			nn.FieldKey = key
-			nn.FieldKeys = ni.FieldValue.MapKeys()
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeMap(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		if IsNilOrInvalidValue(v) {
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: []string{schema.Name},
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for _, key := range ni.FieldValue.MapKeys() {
+				nn := *ni
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = []string{schema.Name}
+				nn.FieldValue = ni.FieldValue.MapIndex(key)
+				nn.FieldKey = key
+				nn.FieldKeys = ni.FieldValue.MapKeys()
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 	}
 
-	return nil
+	return errs
+}
+
+// GetNodes returns the nodes in the data tree at the indicated path, relative
+// to the supplied root. If for a list node is missing, all values in the list
+// are returned. If the key is partial, all nodes matching the values present
+// in the key are returned.
+// If the root is the tree root, the path may be absolute.
+// GetNodes returns an error if the path is not found in the tree, or an element
+// along the path is nil.
+func GetNodes(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	return getNodesInternal(schema, root, path)
+}
+
+// getNodesInternal is the internal implementation of GetNode. In addition to
+// GetNode functionality, it can accept non GoStruct types e.g. map for a keyed
+// list, or a leaf.
+func getNodesInternal(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	if IsValueNil(root) {
+		ResetIndent()
+		return nil, nil, nil
+	}
+	if len(path.GetElem()) == 0 {
+		ResetIndent()
+		return []interface{}{root}, []*yang.Entry{schema}, nil
+	}
+	if schema == nil {
+		return nil, nil, fmt.Errorf("nil schema for data element type %T, remaining path %v", root, path)
+	}
+	// Strip off the absolute path prefix since the relative and absolute paths
+	// are assumed to be equal.
```

Comment:
I don't think this is correct. According to the path specification the root node is encoded as the empty set of elements. Instead, this should be an error if the name of the first element is "".

Reply:
(NB: goyang's paths don't meet the requirements of the `PathElem` gNMI paths, but they shouldn't ever be supplied as `[]string`, and the way that they schema tree is rebuilt in the supplied code, I don't think that you should ever see `[]string{"", "root-elem"}`.

Reply:
This is something that I find confusing in string to gnmi path conversion (the paths here may be coming from leafref paths that may be absolute). What does /a/b/c vs. a/b/c convert to? If they are the same, how can we recognize an absolute path?

Reply:
After our discussion and sleeping on this a bit I'm wondering if we should be insisting that gnmi path spec be followed rigidly here (because we are using it for convenience, rather than in its original domain). Maybe we could state somewhere that we expect both absolute and relative paths as gpb.Path and the former start with ""?

Reply:
I just wonder whether this will cause confusion elsewhere -- since for `GetNode` etc. you'd expect that these paths that are prefixed with "" are not valid, I think. I'm fine with not requiring the path specification to be followed rigidly, if there's a good reason to do this and we can make it clear that these paths are only internal.

Looking at this again, I wasn't (and probably forgot from our previous discussion) entirely clear where the absolute vs. relative case is interesting in this fn?

Reply:
We are converting a leafref path to gnmi and this can be absolute or relative. Since absolute == relative if root is the tree base the code/comment here is just saying if you're passing in an absolute path, I hope you are passing the root element along with it.